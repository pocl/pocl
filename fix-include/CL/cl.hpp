/* cl.hpp - fix for the Khronos Group header

   Copyright (c) 2012 Vincent Danjean <Vincent.Danjean@ens-lyon.org>

   Permission is hereby granted, free of charge, to any person obtaining a copy
   of this software and associated documentation files (the "Software"), to deal
   in the Software without restriction, including without limitation the rights
   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
   copies of the Software, and to permit persons to whom the Software is
   furnished to do so, subject to the following conditions:

   The above copyright notice and this permission notice shall be included in
   all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
   THE SOFTWARE.
*/

// fix problems with current khronos cl.hpp header file
// khronos cl.hpp is a 1.1 OpenCL header file so it needs 1.1 declaration
// defining CL_USE_DEPRECATED_OPENCL_1_1_APIS allow the compilation
// however, this generates warnings
// pragmas are here to avoid the warnings
// but g++ has a bug for now: http://bug.debian.org/686178

#define CL_USE_DEPRECATED_OPENCL_1_1_APIS

#if defined __clang__

#  pragma clang diagnostic push
#    pragma clang diagnostic ignored "-W#warnings"
#    include_next <CL/cl.hpp>
#  pragma clang diagnostic pop

#elif defined GCC_VERSION && GCC_VERSION >= 40600

#  pragma GCC diagnostic push
#    pragma GCC diagnostic ignored "-Wcpp"
#    include_next <CL/cl.hpp>
#  pragma GCC diagnostic pop

#else

#  include_next <CL/cl.hpp>

#endif
