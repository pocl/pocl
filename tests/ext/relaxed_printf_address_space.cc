/* relaxed_printf_address_space.cc -
   Test cases for cl_ext_relaxed_printf_address_space

   Copyright (c) 2022 Pekka Jääskeläinen / Parmance

   Permission is hereby granted, free of charge, to any person obtaining a copy
   of this software and associated documentation files (the "Software"), to
   deal in the Software without restriction, including without limitation the
   rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
   sell copies of the Software, and to permit persons to whom the Software is
   furnished to do so, subject to the following conditions:

   The above copyright notice and this permission notice shall be included in
   all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
   IN THE SOFTWARE.
*/

#include <CL/cl.h>

#include <vector>
#include <fstream>
#include <sstream>
#include <iostream>
#include <string.h>

#define CHECK_ERR(X) do {					\
    if ((X) != CL_SUCCESS) {					\
      std::cerr << #X "ERROR at " << __LINE__ << std::endl;	\
      abort();							\
    } } while (0)

std::vector<std::string> getKernelNames(cl_program Program) {
  size_t KernelNamesSize;
  CHECK_ERR(clGetProgramInfo(Program, CL_PROGRAM_KERNEL_NAMES,
			     0, NULL,
			     &KernelNamesSize));


  char Names[KernelNamesSize + 1];
  CHECK_ERR(clGetProgramInfo(Program, CL_PROGRAM_KERNEL_NAMES,
			     KernelNamesSize, &Names[0],
			     &KernelNamesSize));

  // The specs doesn't ensure a null terminated string.
  Names[KernelNamesSize] = 0;

  std::vector<std::string> KernelNames;
  //Construct a stream from the string
  std::stringstream StreamData(Names);
  std::string Kern;
  while (std::getline(StreamData, Kern, ';')) {
    KernelNames.push_back(Kern);
  }
  return KernelNames;
}

int main(int Argc, char *Argv[]) {

  cl_uint Count;
  cl_platform_id Platform;

  clGetPlatformIDs (1, &Platform, &Count);
  if (Count == 0)
    abort();

  cl_context_properties Properties[] =
    {CL_CONTEXT_PLATFORM, (cl_context_properties)Platform, 0};

  cl_context Context =
    clCreateContextFromType (Properties, CL_DEVICE_TYPE_ALL,
			     NULL, NULL, NULL);

  cl_uint NumDevices;
  cl_int Err = clGetDeviceIDs(Platform, CL_DEVICE_TYPE_ALL, 0, NULL,
			      &NumDevices);
  if (Err != CL_SUCCESS || NumDevices == 0)
    return EXIT_FAILURE;

  cl_device_id Device;
  Err = clGetDeviceIDs (Platform, CL_DEVICE_TYPE_ALL, 1, &Device, NULL);
  if (Err != CL_SUCCESS)
    return EXIT_FAILURE;

  std::ifstream File(SRCDIR "/relaxed_printf_address_space.cl",
		     std::ios::binary | std::ios::ate);
  std::streamsize SrcSize = File.tellg();
  File.seekg(0, std::ios::beg);

  std::vector<char> Src(SrcSize);
  if (!File.read(Src.data(), SrcSize))
    return EXIT_FAILURE;

  const char *SrcCStr = (const char*)Src.data();

  cl_program Program =
    clCreateProgramWithSource (Context, 1, &SrcCStr,
			       (const size_t*)&SrcSize, &Err);
  CHECK_ERR(Err);

  Err = clBuildProgram (Program, 0, NULL, "", NULL, NULL);
  CHECK_ERR(Err);

  cl_command_queue Queue =
    clCreateCommandQueueWithProperties(Context, Device, 0, &Err);
  CHECK_ERR(Err);

  cl_kernel Kernel = clCreateKernel(Program,
				    "relaxed_printf_address_space_tests",
				    NULL);
  int IO = 1;
  cl_mem IOBuffer =
    clCreateBuffer (Context,
		    CL_MEM_READ_WRITE | CL_MEM_COPY_HOST_PTR,
		    sizeof(int), &IO, &Err);
  CHECK_ERR(Err);
  Err = clSetKernelArg (Kernel, 0, sizeof (cl_mem), (void *)&IOBuffer);
  CHECK_ERR(Err);

  char GlobalFmtStr[] = "%s through a global format str\n\0";
  cl_mem GlobalFmtStrBuffer =
    clCreateBuffer (Context,
		    CL_MEM_READ_WRITE | CL_MEM_COPY_HOST_PTR,
		    strlen(GlobalFmtStr) + 1, &GlobalFmtStr, &Err);
  CHECK_ERR(Err);
  Err = clSetKernelArg (Kernel, 1, sizeof (cl_mem),
			(void *)&GlobalFmtStrBuffer);
  CHECK_ERR(Err);

  char GlobalStr[] = "global address space str arg\0";
  cl_mem GlobalStrBuffer =
    clCreateBuffer (Context,
		    CL_MEM_READ_WRITE | CL_MEM_COPY_HOST_PTR,
		    strlen(GlobalStr) + 1, &GlobalStr, &Err);
  CHECK_ERR(Err);
  Err = clSetKernelArg (Kernel, 2, sizeof (cl_mem),
			(void *)&GlobalStrBuffer);
  CHECK_ERR(Err);

  Err = clSetKernelArg (Kernel, 3, 1000, NULL);
  CHECK_ERR(Err);

  Err = clSetKernelArg (Kernel, 4, 1000, NULL);
  CHECK_ERR(Err);

  const size_t LocalSize[] = {1, 1, 1};
  const size_t GlobalSize[] = {1, 1, 1};

  Err = clEnqueueNDRangeKernel (Queue, Kernel, 1, NULL, LocalSize,
				GlobalSize, 0, NULL, NULL);
  CHECK_ERR(Err);

  Err = clEnqueueReadBuffer (Queue, IOBuffer, CL_TRUE, 0,
			       sizeof(int), &IO, 0, NULL, NULL);
  CHECK_ERR(Err);

  clFinish(Queue);

  std::cout << "IO == " << IO << std::endl << std::endl;

  return EXIT_SUCCESS;
}
