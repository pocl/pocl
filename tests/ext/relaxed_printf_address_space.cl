/* relaxed_printf_address_space.cl -
   Test cases for cl_ext_relaxed_printf_address_space

   Copyright (c) 2022 Pekka Jääskeläinen / Parmance

   Permission is hereby granted, free of charge, to any person obtaining a copy
   of this software and associated documentation files (the "Software"), to
   deal in the Software without restriction, including without limitation the
   rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
   sell copies of the Software, and to permit persons to whom the Software is
   furnished to do so, subject to the following conditions:

   The above copyright notice and this permission notice shall be included in
   all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
   IN THE SOFTWARE.
*/

#ifndef cl_ext_relaxed_printf_address_space
#error cl_ext_relaxed_printf_address_space not supported by the device!
#endif

constant const char *constant c_fmt = "%s through a constant format str\n";
constant const char *constant c_arg = "constant address space str arg";

constant const char *constant l_fmt_init = "%s through a local format str\n";
constant const char *constant l_arg_init = "local address space str arg";

constant const char *constant p_fmt_init = "%s through a private format str\n";
constant const char *constant p_arg_init = "private address space str arg";

#define strcpy(dst, src) do {			\
    char c;					\
    int i = 0;					\
    do {					\
      c = src[i];				\
      dst[i] = c;				\
      ++i;					\
    } while (c != '\0');			\
  } while (0)

kernel void relaxed_printf_address_space_tests(global int *io,
					       global char *g_fmt,
					       global char *g_arg,
					       local char *l_fmt,
					       local char *l_arg) {
  char p_arg[1000];
  char p_fmt[1000];

  strcpy(l_fmt, l_fmt_init);
  strcpy(l_arg, l_arg_init);

  strcpy(p_arg, p_arg_init);
  strcpy(p_fmt, p_fmt_init);

  io[0] = printf(g_fmt, g_arg);
  io[0] += printf(c_fmt, c_arg);
  io[0] += printf(l_fmt, l_arg);
  io[0] += printf(p_fmt, p_arg);

  printf("\n");

  io[0] += printf(g_fmt, p_arg);
  io[0] += printf(g_fmt, c_arg);
  io[0] += printf(g_fmt, l_arg);
  io[0] += printf(g_fmt, g_arg);

  printf("\n");

  io[0] += printf(c_fmt, p_arg);
  io[0] += printf(c_fmt, g_arg);
  io[0] += printf(c_fmt, l_arg);
  io[0] += printf(c_fmt, c_arg);

  printf("\n");

  io[0] += printf(l_fmt, p_arg);
  io[0] += printf(l_fmt, g_arg);
  io[0] += printf(l_fmt, c_arg);
  io[0] += printf(l_fmt, l_arg);

  printf("\n");

  io[0] += printf(p_fmt, l_arg);
  io[0] += printf(p_fmt, g_arg);
  io[0] += printf(p_fmt, c_arg);
  io[0] += printf(p_fmt, p_arg);
}
