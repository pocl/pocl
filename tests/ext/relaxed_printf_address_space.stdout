global address space str arg through a global format str
constant address space str arg through a constant format str
local address space str arg through a local format str
private address space str arg through a private format str

private address space str arg through a global format str
constant address space str arg through a global format str
local address space str arg through a global format str
global address space str arg through a global format str

private address space str arg through a constant format str
global address space str arg through a constant format str
local address space str arg through a constant format str
constant address space str arg through a constant format str

private address space str arg through a local format str
global address space str arg through a local format str
constant address space str arg through a local format str
local address space str arg through a local format str

local address space str arg through a private format str
global address space str arg through a private format str
constant address space str arg through a private format str
private address space str arg through a private format str
IO == 0

